### PR TITLE
DStruct for loops upgraded to foreachs

### PR DESCRIPTION
**Pull request by dukc**
My first pull request for the compiler itself. I refactored dstruct.d loops, kind of a practice pull request.

### CODE REVIEW

**Comment by dukc** [comment]
Something went wrong, yet I think it compiled when I tried myself. What's the correct way to test one's changes locally?

File: src/ddmd/dstruct.d [hunk 1]
```diff
@@ -339,22 +339,19 @@ extern (C++) class StructDeclaration : AggregateDeclaration
         /* Set scope so if there are forward references, we still might be able to
          * resolve individual members like enums.
          */
-        for (size_t i = 0; i < members.dim; i++)
+        foreach(s; (*members)[])
```
Comment (yebblies) [inline src/ddmd/dstruct.d]: The `[]` shouldn't be necessary.

File: src/ddmd/dstruct.d [hunk 2]
```diff
@@ -619,9 +616,8 @@ extern (C++) class StructDeclaration : AggregateDeclaration
 
         size_t nfields = fields.dim - isNested();
         size_t offset = 0;
-        for (size_t i = 0; i < elements.dim; i++)
+        foreach(i, e; (*elements)[])
         {
-            Expression e = (*elements)[i];
```
Comment (yebblies) [inline src/ddmd/dstruct.d]: The segfault might be because `e` is reassigned below.  Try it without this part.

**Review by yebblies — CHANGES_REQUESTED** [review]

**Comment by dukc** [comment]
Ah, it was a runtime error, not a compilation error. That explains why I did not notice an error locally.

File: src/ddmd/dstruct.d [hunk 3]
```diff
@@ -339,22 +339,19 @@ extern (C++) class StructDeclaration : AggregateDeclaration
         /* Set scope so if there are forward references, we still might be able to
          * resolve individual members like enums.
          */
-        for (size_t i = 0; i < members.dim; i++)
+        foreach(s; (*members))
```
Comment (yebblies) [inline src/ddmd/dstruct.d]: The `()` shouldn't be necessary either now.

**Comment by dukc** [comment]
True, but above all I have to hunt that segfault out. Is there any better way to test the freshly-made compiler than setting it up for command-line use? That would get slow and error-prone.

**Comment by yebblies** [comment]
I suggest enabling these changes one by one until you find which loop is failing, then sticking in thousands of asserts etc to see if the 'this' pointer is null, the struct pointer is null, the struct is corrupt, the array from opSlice is corrupt etc.

**Comment by dukc** [comment]
That's what I'm gonna do, but what I'm wondering is that how do test a compiler I've made locally? If an autotester can do it, I suppose there's a better way than replacing the "system" dmd with it and testing manually?

**Comment by JackStouffer** [comment]
You can run `make -f posix.mak test`

**Comment by dukc** [comment]
Thanks, that helped. `make -fwin32.mak auto-tester-build` appears to do what i want. But the tests do not compile because phobos templates do not yet work with -dip1000. Should I try to tweak it to compile the test build without that flag? No quarantees about success, I'm not used to makefiles.

**Comment by WalterBright** [comment]
Unfortunately, making this change is not reliable, because sometimes down in the bowels of the code members can be added to `members[]` inside the loop. The `foreach` assumes that `members[]` does not change.

**Comment by dukc** [comment]
What about defining opApply template for Array!T so that it behaves just like the for loop?

**Comment by UplinkCoder** [comment]
churn ...

**Comment by dukc** [comment]
I added a function to the Array!T which should solve that concern of Bright. Unfortunately I still do not know how to run the auto-test locally, so we're lucky if this passes.

**Comment by UplinkCoder** [comment]
This is useless bloat.

**Comment by dukc** [comment]
I wouldn't say useless, because it lets us to do those foreachs instead of fencepost-bug-prone for loops. But whether it's useful enough to justify the weight... you decide.

Do you want it closed, or just simplified somehow? Other opinions welcome too.

**Comment by dukc** [comment]
Plausible opinion. If others agree we don't want it, I will close.

**Comment by WalterBright** [comment]
Pulling this is likely to cause strange bugs in the future, because as I said, because of lazy evaluation of things, sometimes members can be added while executing the loop. The foreach assumes that no members are added, hence strange bugs. We've had these problems several times before with foreach over members, which is why it is not done anymore.

**Pull request closed without merging**